mono/utils/os-event.h

```diff
@@ -0,0 +1,62 @@
+
+#ifndef _MONO_UTILS_OS_EVENT_H_
+#define _MONO_UTILS_OS_EVENT_H_
+
+#include <config.h>
+#include <glib.h>
+
+#include "mono-os-mutex.h"
+
+#ifndef MONO_INFINITE_WAIT
+#define MONO_INFINITE_WAIT ((guint32) 0xFFFFFFFF)
+#endif
+
+#define MONO_OS_EVENT_WAIT_MAXIMUM_OBJECTS 64
+
+typedef enum {
+	MONO_OS_EVENT_WAIT_RET_SUCCESS_0 =  0,
+	MONO_OS_EVENT_WAIT_RET_ALERTED   = -1,
+	MONO_OS_EVENT_WAIT_RET_TIMEOUT   = -2,
+} MonoOSEventWaitRet;
+
+typedef struct _MonoOSEvent MonoOSEvent;
+
+typedef void (*MonoOSEventFreeCb) (MonoOSEvent*);
+
+struct _MonoOSEvent {
+#ifdef HOST_WIN32
+	gpointer handle;
+#else
+	mono_mutex_t mutex;
+	mono_cond_t cond;
+	gboolean manual;
+	gboolean signalled;
+	guint32 set_count;
+#endif
+
+	guint32 ref;
+	MonoOSEventFreeCb free_cb;
+};
+
+void
+mono_os_event_init (MonoOSEvent *event, gboolean manual, gboolean initial, MonoOSEventFreeCb free_cb);
```

Comment:
If you don't use the lifetime management that's perfectly fine, just init it, and unref it when you want to destroy it. I started by having a `MonoOSEvent` and a `MonoOSRefEvent` but that turned out more confusing than having the ref/unref on `MonoOSEvent`.

Like you suggests, adding documentation will definitely help to understand how to use it.